test/unit_tests/oc/hoc_interpreter.cpp

```diff
@@ -11,3 +11,36 @@ TEST_CASE("Test hoc interpreter", "[Neuron][hoc_interpreter]") {
     hoc_add();
     REQUIRE(hoc_xpop() == 9.0);
 }
+
+SCENARIO("Test small HOC functions", "[NEURON][hoc_interpreter]") {
+    // This is targeting errors in ModelDB 136095 with #1995
+    GIVEN("An objref that gets redeclared with fewer indices") {
+        REQUIRE(hoc_oc("objref ncl[1][1][1]\n"
+                       "obfunc foo() { localobj x\n"
+                       "  return ncl\n"
+                       "}\n"
+                       "objref ncl\n"
+                       "foo()\n") == 0);
```

Comment:
@nrnhines can you comment regarding whether this is supposed to work?

It gives an (unhelpful and indirect) error with this PR (https://github.com/neuronsimulator/nrn/pull/1995#issuecomment-1265420120) but not with master.

I **believe** that on master the underlying issue is still there (doubles get pushed on to the stack for the indices and they never get popped), but it goes undiagnosed because https://github.com/neuronsimulator/nrn/blob/0cdcc5c74649686d2c512a297d86e870ab6aac73/src/oc/code.cpp#L1491 does not check that `obj` is active.

Reply:
I expect it to return a NullObject but it wouldn't surprise me if the redeclaration of ncl from array of objects to a scalar object causes problems due to an internal bug. Note that in HOC, leaving off the indices for an array, returns the first element of the array.
```
oc>objref ncl[1][1][1]
oc>ncl
	NULLobject 
oc>ncl[0]
nrniv: wrong # of subscripts ncl
 near line 12
 ncl[0]
       ^
oc>ncl[0][0][0]
	NULLobject 
oc>
```
Anyway,
```
oc>objref ncl[1][1][1]
oc>obfunc foo() {localobj x  return ncl}
oc>foo()
	NULLobject 
oc>objref ncl
oc>foo()
	NULLobject 
oc>
```
is expected but if the obfunc used ```return ncl[0][0][0]```. Then redeclaring it to something without 3 dimensions should raise an error but it does not!
```
oc>objref ncl[1][1][1]
oc>obfunc foo() {localobj x  return ncl[0][0][0]}
oc>foo()
	NULLobject 
oc>objref ncl
oc>foo()
	NULLobject 
oc>
```
but then
```
oc>objref ncl[1]
oc>foo()
	NULLobject 
oc>objref ncl[1][1][1][1]
oc>foo()
bad stack access: expecting (double); really (Unknown)
nrniv: interpreter stack type error
 near line 33
 foo()
      ^
        foo()
oc>
```
This reminds me of problems on the python side I've been trying to generate error messages for.  I.e when a hoc array declaration changes but on the python side one has python objects that refer to partially filled out versions of the old hoc array. I.e.
```
h('objref ncl[2][2][2]')
a = h.ncl[1][1]
print(a[0])
```

Reply:
OK, I think the issue that has become an error is that in your example
```
oc>objref ncl[1][1][1]
oc>obfunc foo() {localobj x  return ncl}
oc>foo()
	NULLobject 
oc>objref ncl
oc>foo()
	NULLobject 
oc>
```
then even in master the second call to `foo()` does not clean up after itself properly.
Just after `hoc_call` for the first `foo()` then `master` has:
```
(lldb) print (void)hoc_prstack()
interpreter stack: 1
0 stacktype=8
```
while just after `hoc_call` for the second `foo()` then `master` has:
```
(lldb) print (void)hoc_prstack()
interpreter stack: 4
0 stacktype=8
1 stacktype=259
2 stacktype=259
3 stacktype=1
```
259 is `NUMBER`, 8 is `OBJECTTMP`, 1 is `USERINT`.
Without wanting to write too much of an essay here, I think that three NUMBER are pushed for the indices and never freed, and some other logic that expects to be popping other things actually pops some of those NUMBERs.

Reply:
I agree. The problem seems to be that when the stack machine for foo is created the correct number of stack slots are reserved for the array variable. But when foo is called, there is no check that the current incarnation of that variable has the same number of indices as were reserved.

Reply:
I opened https://github.com/neuronsimulator/nrn/pull/2009 with a test that fails on ~master (OK, I added an extra check to get it to actually fail).

Reply:
What do you think of the following proposal? For context, consider that
```
oc>objref a[2][3][4]
oc>debug
oc>objref a[2][3][4]
0x140008000 0x1030f5390 CONSTPUSH
constant or string pointer
0x140008010 0x1030f5390 CONSTPUSH
constant or string pointer
0x140008020 0x1030f5390 CONSTPUSH
constant or string pointer
0x140008030 0x1031082e8 objvardecl
a
relative 3
STOP
oc>obfunc foo(){ return a }
0x140008000 0x1030f53cc PUSHZERO
0x140008008 0x1030f53cc PUSHZERO
0x140008010 0x1030f53cc PUSHZERO
0x140008018 0x10310860c objectvar
a
0x140008028 0x10310927c objecteval
0x140008030 0x1030f7ce0 HOCOBJRET
0x140008038 0x1030f662c PROCRET
oc>
```
The proposal is that when the stack machine for foo is constructed, the number of indexes pushed onto the stack  (in this case '3') is put onto the machine after the  ```a``` Symbol. Then, when foo is called, when the machine gets to ```objectvar``` and that isstruction reads the next instruction ```a```, then the following instruction, ```3```, could be compared to the dimension of ```a``` and an error raised if the number of indices don't match.  A number of implementation variants are possible (maybe it should be done by ```objecteval```, but it is a shame to have to pay the overhead of checking dimension consistency even for scalars. (and other questions arise, such as "are index bounds being checked"? On the Python side, I was toying with the idea of an array_change_count field for Hoc Symbols that could be checked against a python intermediate or _ref_ const copy?))

Reply:
This sounds reasonable, but I think I lack the overview to make a **very** informed opinion!

If I understand correctly, this would improve the error reporting without making the `ncl` example work as-is, so we can and should proceed to merge https://github.com/ModelDBRepository/136095/pull/2 and the PR we are commenting on as-is?

Reply:
> improve the error reporting without making the ncl example work as-is

Correct. All previous versions have this stack issue when an array is re-declared with fewer indices and a not very forthcoming stack error when re-declared with more indices. I think a fix should at least raise a meaningful error if the expected number of indices don't match the actual number of indices. And perhaps go further to raise an error if the index bounds change. But that can be a future PR.

Reply:
This will be followed up in https://github.com/neuronsimulator/nrn/issues/2010.